src/dawn/engine_dawn.cc

```diff
@@ -14,15 +14,143 @@
+                             uint32_t* row_stride_ptr,
+                             uint32_t* size_ptr) {
+  ::dawn::BufferDescriptor descriptor;
+  // TODO(dneto): Handle other formats.
+  if (format != ::dawn::TextureFormat::R8G8B8A8Unorm) {
+    Result("Dawn::MakeFramebufferBuffer: Unhandled framebuffer format");
+  }
+  // Number of bytes for each texel in the default format.
+  const uint32_t texel_size = 4;
+
+  uint32_t row_stride = texel_size * kFramebufferWidth;
+  {
+    // Round up the stride to the minimum image row pitch.
+    const uint32_t spillover = row_stride % kImageRowPitch;
+    if (spillover > 0)
+      row_stride += (kImageRowPitch - spillover);
+    assert(0 == (row_stride % kImageRowPitch));
+  }
+
+  descriptor.size = row_stride * kFramebufferHeight;
+  descriptor.usage =
+      ::dawn::BufferUsageBit::TransferDst | ::dawn::BufferUsageBit::MapRead;
+  *result_ptr = device.CreateBuffer(&descriptor);
+  *row_stride_ptr = row_stride;
+  *size_ptr = descriptor.size;
+  return {};
+}
+
+// Result status object and data pointer resulting from a buffer mapping.
+struct MapResult {
+  Result result;
+  const void* data = nullptr;
+};
+
+// Handles the update from an asynchronous buffer map request, updating the
+// state of the MapResult object hidden inside the |userdata| parameter.
+// On a successfulmapping outcome, set the data pointer in the map result.
+// Otherwise set the map result object to an error.
+void HandleBufferMapCallback(dawnBufferMapAsyncStatus status,
+                             const void* data,
+                             dawnCallbackUserdata userdata) {
+  MapResult& map_result = *reinterpret_cast<MapResult*>(userdata);
+  switch (status) {
+    case DAWN_BUFFER_MAP_ASYNC_STATUS_SUCCESS:
+      map_result.data = data;
+      break;
+    case DAWN_BUFFER_MAP_ASYNC_STATUS_ERROR:
+      map_result.result = Result{"Buffer map for reading failed: error"};
+      break;
+    case DAWN_BUFFER_MAP_ASYNC_STATUS_UNKNOWN:
+    case DAWN_BUFFER_MAP_ASYNC_STATUS_FORCE32:
+      map_result.result = Result{"Buffer map for reading failed: unknown"};
+      break;
+    case DAWN_BUFFER_MAP_ASYNC_STATUS_CONTEXT_LOST:
+      map_result.result = Result{"Buffer map for reading failed: context lost"};
+      break;
+  }
+}
+
+// Maps the given buffer and sends its data pointer back through |data_ptr|.
+// Assumes the buffer has usage bit dawn::BufferUsageBit::MapRead set.
+// Returns a result code.  This can fail if the context is lost, for example.
+MapResult MapBuffer(const ::dawn::Device& device,
+                    const ::dawn::Buffer& buf,
+                    uint32_t size) {
+  MapResult map_result;
+  buf.MapReadAsync(0, size, HandleBufferMapCallback,
+                   static_cast<dawnCallbackUserdata>(
+                       reinterpret_cast<uintptr_t>(&map_result)));
+  device.Tick();
+  // Wait until the callback has been processed.
+  unsigned const max_iters = 100;
```

Comment:
The style guide says we should use int in this scenario

https://google.github.io/styleguide/cppguide.html#Integer_Types 

I don't see the value of using a particularly short specified-width type:

> You should not use the unsigned integer types such as uint32_t, unless there is a valid reason such as representing a bit pattern rather than a number, or you need defined overflow modulo 2^N. In particular, do not use unsigned types to say a number will never be negative. Instead, use assertions for this.